### PR TITLE
[quant] Add quantized Embedding module

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44217 [quant] Move EmbeddingBag eager quantization to static
* **#44208 [quant] Add quantized Embedding module**
* #44207 [quant] Support quantization of embedding lookup operators

Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23547384](https://our.internmc.facebook.com/intern/diff/D23547384)

### CODE REVIEW

File: torch/nn/quantized/modules/embedding.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+class EmbeddingPackedParams(torch.nn.Module):
+    _version = 1
+
+    def __init__(self, num_embeddings, embedding_dim, dtype=torch.quint8):
+        super(EmbeddingPackedParams, self).__init__()
+        self.dtype = dtype
+        if self.dtype == torch.quint8:
+            scales = torch.ones(num_embeddings, dtype=torch.float)
+            zero_points = torch.ones(num_embeddings, dtype=torch.float)
```
Comment (vkuzo) [inline torch/nn/quantized/modules/embedding.py:17]: can we add a comment on why this is initialized to one?

File: torch/nn/quantized/modules/embedding.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                 max_norm: Optional[float] = None, norm_type: float = 2., scale_grad_by_freq: bool = False,
+                 sparse: bool = False, _weight: Optional[Tensor] = None, dtype=torch.quint8) -> None:
+        super(Embedding, self).__init__()
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.sparse = sparse
+
+        if _weight is None:
+            scales = torch.ones(num_embeddings, dtype=torch.float)
+            zero_points = torch.ones(num_embeddings, dtype=torch.float)
```
Comment (vkuzo) [inline torch/nn/quantized/modules/embedding.py:104]: seems like this is the same code as in packed params, perhaps we can only do it once?

**Review by vkuzo — APPROVED** [review]

File: torch/nn/quantized/modules/embedding.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                                                           axis=0, dtype=torch.quint8)
+            self.set_weight(wq)
+        else:
+            raise RuntimeError('Unsupported dtype on quantized embedding!')
+
+    @torch.jit.export
+    def set_weight(self, weight):
+        # type: (torch.Tensor) -> None
+        if self.dtype == torch.quint8:
+            self._packed_weight = torch.ops.quantized.embedding_bag_prepack(weight)
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/embedding.py:29]: Do we support per tensor quantization for packed params?

File: torch/nn/quantized/modules/embedding.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    def forward(self, x):
+        return x
+
+    # Version 1
+    #   self
+    #   |--- _packed_weight : Tensor representing weight of EmbeddingPackedParamsBase
+    #   |--- dtype : torch.dtype
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/embedding.py:49]: Should we also have a field for bitwidth?

File: torch/nn/quantized/modules/embedding.py (hunk 3 above)
Comment (supriyar) [reply torch/nn/quantized/modules/embedding.py:29]: Not atm, we only have per-row quantization support with float qparams.

File: torch/nn/quantized/modules/embedding.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    _FLOAT_MODULE = nn.Embedding
+    _version = 1
+
+    def __init__(self, num_embeddings: int, embedding_dim: int, padding_idx: Optional[int] = None,
+                 max_norm: Optional[float] = None, norm_type: float = 2., scale_grad_by_freq: bool = False,
+                 sparse: bool = False, _weight: Optional[Tensor] = None, dtype=torch.quint8) -> None:
+        super(Embedding, self).__init__()
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.sparse = sparse
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/embedding.py:100]: For my understanding: What does self.sparse do ?

File: torch/nn/quantized/modules/embedding.py (hunk 4 above)
Comment (supriyar) [reply torch/nn/quantized/modules/embedding.py:49]: we can use the tensor dtype to determine the bitwidth, right? Currently it only supports 8bit, but once we add 4-bit  qtensor it should be encoded in the dtype.

File: torch/nn/quantized/modules/embedding.py (hunk 5 above)
Comment (supriyar) [reply torch/nn/quantized/modules/embedding.py:100]: Currently it doesn't do anything for the quantized module. I'll remove it from here.
It is used in the float module for sparse gradients for weight tensor.
